hermes/probe/monitor.go

```diff
@@ -12,9 +12,162 @@
+// a storage system.
+
+package probe
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/google/cloudprober/logger"
+	"github.com/google/cloudprober/metrics"
+	"github.com/google/cloudprober/probes/options"
+
+	metricpb "github.com/google/cloudprober/metrics/proto"
+	probepb "github.com/googleinterns/step224-2020/config/proto"
+)
 
 // MonitorProbe holds aggregate information about all probe runs, per-target.
 // It also holds the config and options used to initialise the probe.
-type MonitorProbe struct{}
+type MonitorProbe struct {
+	name    string
+	config  *probepb.HermesProbeDef
+	targets []*probepb.Target
+	opts    *options.Options
+
+	// probeOpLatency is used to record latency values with distinct labels
+	// per probe operation per target.
+	// Recommended usage: append(probeOpLatency[target][probeOp], newMetric)
+	probeOpLatency map[string]map[string][]*metrics.EventMetrics
+	// apiCallLatency is used to record latency values with distinct labels
+	// per API call per target.
+	// Recommended usage: append(apiCallLatency[target][apiCall], newMetric)
+	apiCallLatency map[string]map[string][]*metrics.EventMetrics
+	logger         *logger.Logger
+}
+
+// Init initializes the probe with the given parameters.
+// This is a required method to implement the cloudprober.Probes.Probe interface.
+func (mp *MonitorProbe) Init(name string, opts *options.Options) error {
+	conf, ok := opts.ProbeConf.(*probepb.HermesProbeDef)
+	if !ok {
+		return fmt.Errorf("invalid argument: opts.ProbeConf is not of type *probepb.HermesProbeDef")
+	}
+	mp.name = name
+	mp.config = conf
+	mp.targets = mp.config.GetTargets()
+	mp.opts = opts
+	mp.probeOpLatency = make(map[string]map[string][]*metrics.EventMetrics)
+	mp.apiCallLatency = make(map[string]map[string][]*metrics.EventMetrics)
+	mp.logger = opts.Logger
+
+	return nil
+}
+
+// reportMetrics sends the metrics recorded in the last probe run to Cloudprober.
+// Arguments:
+//	- metricChan: metric channel passed from Cloudprober
+func (mp *MonitorProbe) reportMetrics(metricChan chan<- *metrics.EventMetrics) {
+	for _, target := range mp.probeOpLatency {
+		for _, op := range target {
+			for _, metric := range op {
+				metricChan <- metric
+			}
+		}
+	}
+
+	for _, target := range mp.apiCallLatency {
+		for _, call := range target {
+			for _, metric := range call {
+				metricChan <- metric
+			}
+		}
+	}
+}
+
+// Start runs the probe indefinitely, unless cancelled, at the configured interval.
+// Probe metrics will be sent via the metricChan at the beginning of the next probe run.
+// Arguments:
+//	- ctx: context provided for cancelling probe.
+//	- metricChan: unidirectional channel used for sending metrics to be surfaced.
+func (mp *MonitorProbe) Start(ctx context.Context, metricChan chan<- *metrics.EventMetrics) {
+	probeTicker := time.NewTicker(mp.opts.Interval)
+
+	for {
+		select {
+		case <-ctx.Done():
+			probeTicker.Stop()
+			return
+		case <-probeTicker.C:
+			mp.reportMetrics(metricChan)
```

Comment:
Could you do this the opposite way around, and return the metrics from mp.runProbe(), then immediately report them?

Reply:
+1
It would be better to only report metrics after a successful probe iteration. It's easier to do that in runProbe().
Either there, or do it right after runProbe().

Reply:
I've moved metric reporting into runProbe() as suggested.